pkg/dynamicaccess/controller.go

```diff
@@ -1,54 +1,210 @@
+	historyRef := *historyRootHash
+	if kvsRef.Equal(swarm.ZeroAddress) {
+		err = c.accessLogic.AddPublisher(ctx, kvs, publisher)
 		if err != nil {
-			return swarm.EmptyAddress, err
+			return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+		}
+		kvsRef, err = kvs.Save(ctx)
+		if err != nil {
+			return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+		}
+		err = history.Add(ctx, kvsRef, &now)
+		if err != nil {
+			return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+		}
+		historyRef, err = history.Store(ctx)
+		if err != nil {
+			return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+		}
+	}
+	encryptedRef, err := c.accessLogic.EncryptRef(ctx, kvs, publisher, refrefence)
+	return kvsRef, historyRef, encryptedRef, err
+}
+
+func NewController(ctx context.Context, accessLogic ActLogic, getter storage.Getter, putter storage.Putter) Controller {
+	return &controller{
+		granteeList: nil, //NewGranteeList(ls, ps, swarm.EmptyAddress),
+		accessLogic: accessLogic,
+		getter:      getter,
+		putter:      putter,
+	}
+}
+
+func (c *controller) Grant(ctx context.Context, granteesAddress swarm.Address, grantee *ecdsa.PublicKey) error {
+	return c.granteeList.Add([]*ecdsa.PublicKey{grantee})
+}
+
+func (c *controller) Revoke(ctx context.Context, granteesAddress swarm.Address, grantee *ecdsa.PublicKey) error {
+	if !c.isRevokeFlagged(granteesAddress) {
+		c.setRevokeFlag(granteesAddress, true)
+	}
+	return c.granteeList.Remove([]*ecdsa.PublicKey{grantee})
+}
+
+func (c *controller) Commit(ctx context.Context, granteesAddress swarm.Address, actRootHash swarm.Address, publisher *ecdsa.PublicKey) (swarm.Address, swarm.Address, error) {
+	//HACKreplace mock with real kvs
+	var act kvs.KeyValueStore
+	if c.isRevokeFlagged(granteesAddress) {
+		act = kvsmock.New()
+		c.accessLogic.AddPublisher(ctx, act, publisher)
+	} else {
+		act = kvsmock.NewReference(actRootHash)
+	}
+
+	grantees := c.granteeList.Get()
+	for _, grantee := range grantees {
+		c.accessLogic.AddGrantee(ctx, act, publisher, grantee, nil)
+	}
+
+	//HACK: Store not implemented
+	granteeref, err := c.granteeList.Save(ctx)
+	if err != nil {
+		return swarm.EmptyAddress, swarm.EmptyAddress, err
+	}
+
+	actref, err := act.Save(ctx)
+	if err != nil {
+		return swarm.EmptyAddress, swarm.EmptyAddress, err
+	}
+
+	c.setRevokeFlag(granteesAddress, false)
+	return granteeref, actref, err
+}
+
+func (c *controller) HandleGrantees(ctx context.Context, granteesAddress swarm.Address, publisher *ecdsa.PublicKey, addList, removeList []*ecdsa.PublicKey) error {
+	act := kvsmock.New()
+
+	c.accessLogic.AddPublisher(ctx, act, publisher)
+	for _, grantee := range addList {
+		c.accessLogic.AddGrantee(ctx, act, publisher, grantee, nil)
+	}
+	// granteeList.Store()
+	return nil
+}
+
+func (c *controller) GetGrantees(ctx context.Context, granteeRootHash swarm.Address) ([]*ecdsa.PublicKey, error) {
+	//[ ]: grantee list address deterministic or stored?
```

Comment:
Should this have a TODO prefix to track it? The "grateeListAddress" line after is still relevant as a comment?

Reply:
Gratnee related functions will be handled in a later PR.